src/clr-angular/utils/drag-and-drop/droppable/droppable.ts

```diff
@@ -0,0 +1,205 @@
+            point.pageX <= this.clientRect.right + this._dropTolerance.right &&
+            point.pageY >= this.clientRect.top - this._dropTolerance.top &&
+            point.pageY <= this.clientRect.bottom + this._dropTolerance.bottom) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private onDragStart(dragStartEvent: ClrDragEvent<T>): void {
+        // Check draggable and droppable have a matching group key.
+        this.isDraggableMatch = this.checkGroupMatch(dragStartEvent.group);
+
+        // Subscribe to dragMoved and dragEnded only if draggable and droppable have a matching group key.
+        if (this.isDraggableMatch) {
+            this.dragStartEmitter.emit(dragStartEvent);
+            this.dragMoveSubscription = this.eventBus.dragMoved.subscribe((dragMoveEvent: ClrDragEvent<T>) => {
+                this.onDragMove(dragMoveEvent);
+            });
+            this.dragEndSubscription = this.eventBus.dragEnded.subscribe((dragEndEvent: ClrDragEvent<T>) => {
+                this.onDragEnd(dragEndEvent);
+            });
+        }
+    }
+
+    private onDragMove(dragMoveEvent: ClrDragEvent<T>): void {
+        if (!this._isDraggableOver && this.isInDropArea(dragMoveEvent.dropPointPosition)) {
+            this.isDraggableOver = true;
+            const dragEnterEvent = this.transformAndBroadcastEvent(dragMoveEvent, ClrDragEventType.DRAG_ENTER);
+            this.dragEnterEmitter.emit(dragEnterEvent);
+        } else if (this._isDraggableOver && !this.isInDropArea(dragMoveEvent.dropPointPosition)) {
+            this.isDraggableOver = false;
+            const dragLeaveEvent = this.transformAndBroadcastEvent(dragMoveEvent, ClrDragEventType.DRAG_LEAVE);
+            this.dragLeaveEmitter.emit(dragLeaveEvent);
+        }
+
+        this.dragMoveEmitter.emit(dragMoveEvent);
+    }
+
+    private onDragEnd(dragEndEvent: ClrDragEvent<T>): void {
+        if (this._isDraggableOver) {
+            if (dragEndEvent.ghostElement) {
+                // By this point draggable ghost component would have been destroyed,
+                // but the element would be active until its animation completes.
+                // As such, once the ghost is dropped over, we will give it "dropped" class.
+                this.renderer.addClass(dragEndEvent.ghostElement, "dropped");
+            }
+            const dropEvent = this.transformAndBroadcastEvent(dragEndEvent, ClrDragEventType.DROP);
+            this.dropEmitter.emit(dropEvent);
+            this.isDraggableOver = false;
+        }
+        this.dragEndEmitter.emit(dragEndEvent);
+        this.unsubscribeFrom(this.dragMoveSubscription);
+        this.unsubscribeFrom(this.dragEndSubscription);
+        this.isDraggableMatch = false;
+        delete this.clientRect;
+    }
+
+    private transformAndBroadcastEvent(event: ClrDragEvent<T>, newEventType: ClrDragEventType): ClrDragEvent<T> {
+        event.type = newEventType;
```

Comment:
Wait, you're modifying the event in place! Isn't the event object shared with everyone? I don't know why you don't have bugs with this, maybe I'm missing something.
Anyway, I'm positive you should do:
```typescript
newEvent = {...event, type: newEventType}
this.eventBus.broadcast(newEvent);
```

Reply:
Thanks, I haven't thought of this. I used this method in other areas too :)